src/main/java/org/wildfly/security/auth/provider/FileSystemSecurityRealm.java

```diff
@@ -382,36 +371,33 @@ public void create() throws RealmUnavailableException {
             }
         }
 
-        public void setCredential(final Object credential) throws RealmUnavailableException {
-            Assert.checkNotNullParam("credential", credential);
+        public void deleteCredential(final String credentialName) throws RealmUnavailableException {
+            Assert.checkNotNullParam("credential", credentialName);
             final LoadedIdentity loadedIdentity = loadIdentity(false, false);
             if (loadedIdentity == null) {
                 throw ElytronMessages.log.fileSystemRealmNotFound(name);
             }
-            List<Object> credentials = loadedIdentity.getCredentials();
-            removeCredentialOfTheSameType(credentials, credential);
-            credentials.add(credential);
+            Map<String, Object> credentials = loadedIdentity.getCredentials();
+            credentials.remove(credentialName);
 
             final LoadedIdentity newIdentity = new LoadedIdentity(getName(), credentials, loadedIdentity.getAttributes());
             replaceIdentity(newIdentity);
         }
 
-        private void removeCredentialOfTheSameType(List<Object> credentials, Object pattern) {
-            for (Iterator iterator = credentials.iterator(); iterator.hasNext();) {
-                Object credential = iterator.next();
-                if (credential instanceof Key && pattern instanceof Key) { // keys comparison
-                    if (((Key)credential).getAlgorithm().equals(((Key)pattern).getAlgorithm())) {
-                        iterator.remove();
-                    }
-                } else { // non-keys comparison
-                    if(credential.getClass() == pattern.getClass()) {
-                        iterator.remove();
-                    }
-                }
+        public void setCredential(final String credentialName, final Object credential) throws RealmUnavailableException {
+            Assert.checkNotNullParam("credential", credential);
+            final LoadedIdentity loadedIdentity = loadIdentity(false, false);
+            if (loadedIdentity == null) {
+                throw ElytronMessages.log.fileSystemRealmNotFound(name);
             }
+            Map<String, Object> credentials = loadedIdentity.getCredentials();
+            credentials.put(credentialName, credential);
+
+            final LoadedIdentity newIdentity = new LoadedIdentity(getName(), credentials, loadedIdentity.getAttributes());
+            replaceIdentity(newIdentity);
         }
 
-        public void setCredentials(final List<Object> credentials) throws RealmUnavailableException {
+        public void setCredentials(final Map<String, Object> credentials) throws RealmUnavailableException {
```

Comment:
Since the XML schema is changing, you would normally have to add a new elytron-1_1.xsd file which is a copy of the 1_0 file, with the new item(s) in it.  Then the parser code would be changed to emit and parse the 1.1 namespace, while still allowing the 1.0 namespace.  The configuration parser would also have to be updated to support the 1.1 namespace.

However, since (as far as I know) the 1.0 version of the project was used only in a situation where no XML was parsed or written, we could possibly get away with just updating the 1.0 schema.  It'd be @darranl's call.